src/lib/storage/dictionary_segment.hpp

```diff
@@ -25,55 +25,122 @@ class DictionarySegment : public BaseSegment {
   /**
    * Creates a Dictionary segment from a given value segment.
    */
-  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& base_segment);
+  explicit DictionarySegment(const std::shared_ptr<BaseSegment>& baseSegment) {
+    auto valueSegment = std::static_pointer_cast<ValueSegment<T>>(baseSegment);
+
+    _build_compressed_dictionary(valueSegment->values());
+  }
 
   // SEMINAR INFORMATION: Since most of these methods depend on the template parameter, you will have to implement
   // the DictionarySegment in this file. Replace the method signatures with actual implementations.
 
   // return the value at a certain position. If you want to write efficient operators, back off!
-  AllTypeVariant operator[](const ChunkOffset chunk_offset) const override;
+  AllTypeVariant operator[](const ChunkOffset chunk_offset) const {
+    return static_cast<AllTypeVariant>(_dictionary->at(_attribute_vector->at(chunk_offset)));
+  };
 
   // return the value at a certain position.
-  T get(const size_t chunk_offset) const;
+  T get(const size_t chunk_offset) const {
+      return _dictionary->at(_attribute_vector->at(chunk_offset));
+  };
 
   // dictionary segments are immutable
-  void append(const AllTypeVariant& val) override;
+  void append(const AllTypeVariant& val){
+      throw std::runtime_error("Dictionary segments are immutable. You shall not append anything.");
+  };
 
   // returns an underlying dictionary
-  std::shared_ptr<const std::vector<T>> dictionary() const;
+  std::shared_ptr<const std::vector<T>> dictionary() {
+      return _dictionary;
+  };
 
   // returns an underlying data structure
-  std::shared_ptr<const BaseAttributeVector> attribute_vector() const;
+  std::shared_ptr<std::vector<uint32_t>> attribute_vector() const {
+      return _attribute_vector;
+  };
 
   // return the value represented by a given ValueID
-  const T& value_by_value_id(ValueID value_id) const;
+  const T& value_by_value_id(ValueID value_id) const {
+      return _dictionary->at(value_id);
+  };
 
   // returns the first value ID that refers to a value >= the search value
   // returns INVALID_VALUE_ID if all values are smaller than the search value
-  ValueID lower_bound(T value) const;
+  ValueID lower_bound(T value) const {
+    for(ValueID dictionary_index = ValueID{0}; dictionary_index < _dictionary->size(); ++dictionary_index ) {
+      if(_dictionary->at(dictionary_index) >= value) {
```

Comment:
same as above ...
```suggestion
      if(value_by_id(dictionary_index) >= value) {
```

Reply:
I would not always use the public API of the class for internal access of members if the access is not that much of a hustle and readable as is. And it does not feel right..

Reply:
@m0e33

Reply:
@simonstadlinger in the future, if you think that you always want to do a check when reading a value from the attribute vector, or you always want to validate thing X, or you want to log access metrics, you would have to touch all the places where you don't use the public API.

Reusability is the advantage of a method... why not use it? What would be a disadvantage of my approach?